### PR TITLE
Improves the virtual deduplication algorithm

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

The deduplication algorithm runs once the full dependency tree has been expanded. It helps to generate efficient trees, but it doesn't help the install speed since the traversal occurs before the optimization pass. As a result, some projects take a LOT of time to install when they have hundreds of cross-referencing workspaces.

**How did you fix it?**

This diff changes the deduplication algorithm so that it runs after the virtual packages have been created and their connections have been setup, but before we recurse inside them. If we detect that the dependency set is the exact same as another instance, we discard the newly created package and never recurse inside it.

I'm not 100% certain yet I'm not forgetting some edge case, so I hope the tests will be able to shed some light on this 🙂

### CODE REVIEW

File: packages/yarnpkg-core/sources/Project.ts [hunk 1]
```diff
@@ -1631,10 +1632,14 @@ function applyVirtualResolutionMutations({
     return result;
   };
 
+  let iterationCount = 0;
```
Comment (merceyz) [inline packages/yarnpkg-core/sources/Project.ts]: @arcanis Leftover debug code?

Comment (arcanis) [reply packages/yarnpkg-core/sources/Project.ts]: Yep! Will fix 👍
